### PR TITLE
Shut down gradle at the end of each task

### PR DESCRIPTION
**Pull request by Hixie**
Apparently Gradle leaks memory and it's causing failures.

### CODE REVIEW

File: dev/devicelab/lib/framework/runner.dart [hunk 1]
```diff
@@ -125,3 +127,33 @@ Future<VMIsolateRef> _connectToRunnerIsolate(int vmServicePort) async {
...
+        <String>['a+x', path.join(tempDir.path, gradlewBinaryName)],
+        canFail: true,
+      );
+    }
+    await exec(
+      path.join(tempDir.path, gradlewBinaryName),
+      <String>['--stop'],
+      environment: <String, String>{ 'JAVA_HOME': javaHome },
+      workingDirectory: tempDir.path,
+      canFail: true,
```
Comment (DanTup) [inline dev/devicelab/lib/framework/runner.dart]: If we're allowing this to fail, do we want to log if it does (so if the issue doesn't seem to be fixed, we can check that it wasn't failing here)?

File: dev/devicelab/lib/framework/runner.dart [hunk 2]
```diff
@@ -125,3 +127,33 @@ Future<VMIsolateRef> _connectToRunnerIsolate(int vmServicePort) async {
...
+
+Future<void> cleanupSystem() async {
+  print('\n\nCleaning up system after task...');
+  final String javaHome = await findJavaHome();
+  if (javaHome != null) {
+    print('\nTelling Gradle to shut down (JAVA_HOME=$javaHome)');
+    final String gradlewBinaryName = Platform.isWindows ? 'gradlew.bat' : 'gradlew';
+    final Directory tempDir = Directory.systemTemp.createTempSync('flutter_devicelab_shutdown_gradle.');
+    recursiveCopy(new Directory(path.join(flutterDirectory.path, 'bin', 'cache', 'artifacts', 'gradle_wrapper')), tempDir);
+    copy(new File(path.join(path.join(flutterDirectory.path, 'packages', 'flutter_tools'), 'templates', 'create', 'android.tmpl', 'gradle', 'wrapper', 'gradle-wrapper.properties')), new Directory(path.join(tempDir.path, 'gradle', 'wrapper')));
```
Comment (DanTup) [inline dev/devicelab/lib/framework/runner.dart]: Might be worth a comment explaining why we have to do this (I guess we can't run `gradlew` from an arbitrary folder that doesn't have a properties file?)

File: dev/devicelab/lib/framework/runner.dart [hunk 3]
```diff
@@ -125,3 +127,33 @@ Future<VMIsolateRef> _connectToRunnerIsolate(int vmServicePort) async {
...
+      );
+    }
+    await exec(
+      path.join(tempDir.path, gradlewBinaryName),
+      <String>['--stop'],
+      environment: <String, String>{ 'JAVA_HOME': javaHome },
+      workingDirectory: tempDir.path,
+      canFail: true,
+    );
+    rmTree(tempDir);
```
Comment (DanTup) [inline dev/devicelab/lib/framework/runner.dart]: We've had issues in the past with Windows having mysterious file locks when we try to delete newly-created files in cleanup scripts, not sure if we may need to tolerate errors for this? (If you don't have easy access to Windows and want to check before landing, I can see if I can test this a little on my PC later)

**Review by DanTup — COMMENTED** [review]
Added a couple of minor comments, but otherwise LGTM

File: dev/devicelab/lib/framework/runner.dart (hunk 1 above)
Comment (Hixie) [reply dev/devicelab/lib/framework/runner.dart]: exec() always dumps the exit code, iirc

File: dev/devicelab/lib/framework/runner.dart (hunk 2 above)
Comment (Hixie) [reply dev/devicelab/lib/framework/runner.dart]: will do

File: dev/devicelab/lib/framework/runner.dart (hunk 3 above)
Comment (Hixie) [reply dev/devicelab/lib/framework/runner.dart]: `rmTree` uses `rm` which tolerates errors.

**Comment by Hixie** [comment]
Added comment. Thanks for the review. Will land on green unless there are more comments.

File: dev/devicelab/lib/framework/runner.dart (hunk 1 above)
Comment (DanTup) [reply dev/devicelab/lib/framework/runner.dart]: If I'm reading it correctly, it looks silent when `canFail = true`:

```dart
final int exitCode = await process.exitCode;

if (exitCode != 0 && !canFail)
  fail('Executable failed with exit code $exitCode.');

return exitCode;
```

File: dev/devicelab/lib/framework/runner.dart (hunk 3 above)
Comment (DanTup) [reply dev/devicelab/lib/framework/runner.dart]: Doh! I was looking at old code, it looked like this:

```dart
/// Remove recursively.
void rmTree(FileSystemEntity entity) {
  if (entity.existsSync())
    entity.deleteSync(recursive: true);
}
```

But I see it's changed recently! :)

File: dev/devicelab/lib/framework/runner.dart (hunk 1 above)
Comment (Hixie) [reply dev/devicelab/lib/framework/runner.dart]: It's printed by startProcess; see line 238.

Comment (DanTup) [reply dev/devicelab/lib/framework/runner.dart]: Whoops, I thought that was a framework function. All good, thanks! :-)

**Pull request merged**